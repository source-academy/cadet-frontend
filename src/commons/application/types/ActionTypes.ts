// game Dev Action Types
export const FETCH_TEST_STORIES = 'FETCH_TEST_STORIES';
export const SAVE_USER_STATE = 'SAVE_USER_STATE';
export const SET_GAME_STATE = 'SET_GAME_STATE';

// GroundControl Action Types
export const CHANGE_DATE_ASSESSMENT = 'CHANGE_DATE_ASSESSMENT';
export const DELETE_ASSESSMENT = 'DELETE_ASSESSMENT';
export const PUBLISH_ASSESSMENT = 'PUBLISH_ASSESSMENT';
export const UPLOAD_ASSESSMENT = 'UPLOAD_ASSESSMENT';
