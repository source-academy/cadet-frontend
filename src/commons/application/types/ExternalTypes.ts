/**
 * Defined for displaying an external library.
 * @see Library the definition of a Library at AssessmentTypes.ts in an assessment.
 */
export type External = {
    key: number;
    name: ExternalLibraryName;
    symbols: string[];
};

export enum ExternalLibraryNames {
    NONE = 'NONE',
    RUNES = 'RUNES',
    CURVES = 'CURVES',
    SOUNDS = 'SOUNDS',
    BINARYTREES = 'BINARYTREES',
    PIXNFLIX = 'PIX&FLIX',
    MACHINELEARNING = 'MACHINELEARNING',
    ALL = 'ALL'
}

export type ExternalLibraryName = (typeof ExternalLibraryNames)[keyof typeof ExternalLibraryNames];

export type ExternalLibrary = {
    name: ExternalLibraryName;
    symbols: string[];
};

