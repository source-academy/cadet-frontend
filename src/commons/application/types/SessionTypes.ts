import { Grading, GradingOverview } from '../../../features/grading/GradingTypes';
import { Assessment, AssessmentOverview } from '../../assessment/AssessmentTypes';
import { Notification } from '../../notificationBadge/NotificationBadgeTypes';
import { HistoryHelper } from '../../utils/HistoryHelper';
import { GameState, Role, Story } from '../ApplicationTypes';

import { Announcement } from 'src/components/Announcements'; // TODO: Remove
import { DirectoryData, MaterialData } from 'src/components/material/materialShape'; // TODO: Remove

export const FETCH_ANNOUNCEMENTS = 'FETCH_ANNOUNCEMENTS';
export const FETCH_AUTH = 'FETCH_AUTH';
export const FETCH_ASSESSMENT = 'FETCH_ASSESSMENT';
export const FETCH_ASSESSMENT_OVERVIEWS = 'FETCH_ASSESSMENT_OVERVIEWS';
export const FETCH_GRADING = 'FETCH_GRADING';
export const FETCH_GRADING_OVERVIEWS = 'FETCH_GRADING_OVERVIEWS';
export const LOGIN = 'LOGIN';
export const SET_TOKENS = 'SET_TOKENS';
export const SET_USER = 'SET_USER';
export const SUBMIT_ANSWER = 'SUBMIT_ANSWER';
export const SUBMIT_ASSESSMENT = 'SUBMIT_ASSESSMENT';
export const SUBMIT_GRADING = 'SUBMIT_GRADING';
export const SUBMIT_GRADING_AND_CONTINUE = 'SUBMIT_GRADING_AND_CONTINUE';
export const UNSUBMIT_SUBMISSION = 'UNSUBMIT_SUBMISSION';
export const UPDATE_HISTORY_HELPERS = 'UPDATE_HISTORY_HELPERS';
export const UPDATE_ASSESSMENT_OVERVIEWS = 'UPDATE_ASSESSMENT_OVERVIEWS';
export const UPDATE_ASSESSMENT = 'UPDATE_ASSESSMENT';
export const UPDATE_GRADING_OVERVIEWS = 'UPDATE_GRADING_OVERVIEWS';
export const UPDATE_GRADING = 'UPDATE_GRADING';
export const FETCH_NOTIFICATIONS = 'FETCH_NOTIFICATIONS';
export const ACKNOWLEDGE_NOTIFICATIONS = 'ACKNOWLEDGE_NOTIFICATIONS';
export const UPDATE_NOTIFICATIONS = 'UPDATE_NOTIFICATIONS';
export const NOTIFY_CHATKIT_USERS = 'NOTIFY_CHATKIT_USERS'; // TODO: Remove

export type SessionState = {
  readonly accessToken?: string;
  readonly assessmentOverviews?: AssessmentOverview[];
  readonly assessments: Map<number, Assessment>;
  readonly announcements?: Announcement[]; // TODO: Remove
  readonly grade: number;
  readonly gradingOverviews?: GradingOverview[];
  readonly gradings: Map<number, Grading>;
  readonly group: string | null;
  readonly historyHelper: HistoryHelper;
  readonly materialDirectoryTree: DirectoryData[] | null;
  readonly materialIndex: MaterialData[] | null;
  readonly maxGrade: number;
  readonly maxXp: number;
  readonly refreshToken?: string;
  readonly role?: Role;
  readonly story: Story;
  readonly gameState: GameState;
  readonly name?: string;
  readonly xp: number;
  readonly notifications: Notification[];
};
