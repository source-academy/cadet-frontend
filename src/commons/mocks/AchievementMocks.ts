import {
  AchievementAbility,
  AchievementGoal,
  AchievementItem,
  GoalType
} from '../../features/achievement/AchievementTypes';

export const mockAchievements: AchievementItem[] = [
  {
    id: 0,
    title: 'Rune Master',
    ability: AchievementAbility.CORE,
    isTask: true,
    position: 1,
    prerequisiteIds: [2, 1],
    goalIds: [0],
    cardTileUrl:
      'https://source-academy-assets.s3-ap-southeast-1.amazonaws.com/achievement/card-tile/rune-master-tile.png',
    view: {
      canvasUrl:
        'https://source-academy-assets.s3-ap-southeast-1.amazonaws.com/achievement/canvas/sample-canvas.png',
      description:
        'Lorem ipsum dolor sit amet, consectetur adipiscing elit, sed do eiusmod tempor incididunt ut labore et dolore magna aliqua. Ut enim ad minim veniam, quis nostrud',
      completionText: 'Congratulations! You have completed Rune Master achievement'
    }
  },
  {
    id: 1,
    title: 'Beyond the Second Dimension',
    ability: AchievementAbility.CORE,
    deadline: new Date(2020, 7, 6, 12, 30, 0),
    isTask: false,
    position: 0,
    prerequisiteIds: [],
    goalIds: [1],
    cardTileUrl:
      'https://source-academy-assets.s3-ap-southeast-1.amazonaws.com/achievement/card-tile/btsd-tile.png',
    view: {
      canvasUrl:
        'https://source-academy-assets.s3-ap-southeast-1.amazonaws.com/achievement/canvas/sample-canvas.png',
      description:
        'Lorem ipsum dolor sit amet, consectetur adipiscing elit, sed do eiusmod tempor incididunt ut labore et dolore magna aliqua. Ut enim ad minim veniam, quis nostrud',
      completionText: 'Congratulations! You have completed Beyond the Second Dimension achievement'
    }
  },
  {
    id: 2,
    title: 'Colorful Carpet',
    ability: AchievementAbility.CORE,
    deadline: new Date(2020, 7, 8, 9, 0, 0),
    isTask: false,
    position: 0,
    prerequisiteIds: [],
    goalIds: [2],
    cardTileUrl:
      'https://source-academy-assets.s3-ap-southeast-1.amazonaws.com/achievement/card-tile/colorful-carpet-tile.png',
    view: {
      canvasUrl:
        'https://source-academy-assets.s3-ap-southeast-1.amazonaws.com/achievement/canvas/sample-canvas.png',
      description:
        'Lorem ipsum dolor sit amet, consectetur adipiscing elit, sed do eiusmod tempor incididunt ut labore et dolore magna aliqua. Ut enim ad minim veniam, quis nostrud',
      completionText: 'Congratulations! You have completed Colorful Carpet achievement'
    }
  },
  {
    id: 3,
    title: '',
    ability: AchievementAbility.CORE,
    isTask: false,
    position: 0,
    prerequisiteIds: [],
    goalIds: [],
    cardTileUrl: '',
    view: {
      canvasUrl: '',
      description: '',
      completionText: ''
    }
  },
  {
    id: 4,
    title: 'Curve Wizard',
    ability: AchievementAbility.CORE,
    deadline: new Date(2020, 8, 15, 0, 0, 0),
    isTask: true,
    position: 4,
    prerequisiteIds: [5, 6],
    goalIds: [3],
    cardTileUrl:
      'https://source-academy-assets.s3-ap-southeast-1.amazonaws.com/achievement/card-tile/curve-wizard-tile.png',
    view: {
      canvasUrl:
        'https://source-academy-assets.s3-ap-southeast-1.amazonaws.com/achievement/canvas/sample-canvas.png',
      description:
        'Lorem ipsum dolor sit amet, consectetur adipiscing elit, sed do eiusmod tempor incididunt ut labore et dolore magna aliqua. Ut enim ad minim veniam, quis nostrud',
      completionText: 'Congratulations! You have completed Curve Wizard achievement'
    }
  },
  {
    id: 5,
    title: 'Curve Introduction',
    ability: AchievementAbility.CORE,
    deadline: new Date(2020, 7, 28, 0, 0, 0),
    isTask: false,
    position: 0,
    prerequisiteIds: [],
    goalIds: [4],
    cardTileUrl:
      'https://source-academy-assets.s3-ap-southeast-1.amazonaws.com/achievement/card-tile/curve-introduction-tile.png',
    view: {
      canvasUrl:
        'https://source-academy-assets.s3-ap-southeast-1.amazonaws.com/achievement/canvas/sample-canvas.png',
      description:
        'Lorem ipsum dolor sit amet, consectetur adipiscing elit, sed do eiusmod tempor incididunt ut labore et dolore magna aliqua. Ut enim ad minim veniam, quis nostrud',
      completionText: 'Congratulations! You have completed Curve Introduction achievement'
    }
  },
  {
    id: 6,
    title: 'Curve Manipulation',
    ability: AchievementAbility.CORE,
    deadline: new Date(2020, 8, 5, 0, 0, 0),
    isTask: false,
    position: 0,
    prerequisiteIds: [],
    goalIds: [5],
    cardTileUrl:
      'https://source-academy-assets.s3-ap-southeast-1.amazonaws.com/achievement/card-tile/curve-manipulation-tile.png',
    view: {
      canvasUrl:
        'https://source-academy-assets.s3-ap-southeast-1.amazonaws.com/achievement/canvas/sample-canvas.png',
      description:
        'Lorem ipsum dolor sit amet, consectetur adipiscing elit, sed do eiusmod tempor incididunt ut labore et dolore magna aliqua. Ut enim ad minim veniam, quis nostrud',
      completionText: 'Congratulations! You have completed Curve Manipulation achievement'
    }
  },
  {
    id: 7,
    title: 'The Source-rer',
    ability: AchievementAbility.EFFORT,
    deadline: new Date(2020, 7, 21, 0, 0, 0),
    isTask: true,
    position: 3,
    prerequisiteIds: [],
    goalIds: [6, 7],
    cardTileUrl:
      'https://source-academy-assets.s3-ap-southeast-1.amazonaws.com/achievement/card-tile/the-source-rer-tile.png',
    view: {
      canvasUrl:
        'https://source-academy-assets.s3-ap-southeast-1.amazonaws.com/achievement/canvas/sample-canvas.png',
      description:
        'Lorem ipsum dolor sit amet, consectetur adipiscing elit, sed do eiusmod tempor incididunt ut labore et dolore magna aliqua. Ut enim ad minim veniam, quis nostrud',
      completionText: 'Congratulations! You have completed The Source-rer achievement'
    }
  },
  {
    id: 8,
    title: 'Power of Friendship',
    ability: AchievementAbility.COMMUNITY,
    isTask: true,
    position: 2,
    prerequisiteIds: [9],
    goalIds: [],
    cardTileUrl:
      'https://source-academy-assets.s3-ap-southeast-1.amazonaws.com/achievement/card-tile/power-of-friendship-tile.png',
    view: {
      canvasUrl:
        'https://source-academy-assets.s3-ap-southeast-1.amazonaws.com/achievement/canvas/sample-canvas.png',
      description:
        'Lorem ipsum dolor sit amet, consectetur adipiscing elit, sed do eiusmod tempor incididunt ut labore et dolore magna aliqua. Ut enim ad minim veniam, quis nostrud',
      completionText: 'Congratulations! You have completed Power of Friendship achievement'
    }
  },
  {
    id: 9,
    title: 'Piazza Guru',
    ability: AchievementAbility.COMMUNITY,
    isTask: false,
    position: 0,
    prerequisiteIds: [],
    goalIds: [8],
    cardTileUrl:
      'https://source-academy-assets.s3-ap-southeast-1.amazonaws.com/achievement/card-tile/piazza-guru-tile.png',
    view: {
      canvasUrl:
        'https://source-academy-assets.s3-ap-southeast-1.amazonaws.com/achievement/canvas/sample-canvas.png',
      description:
        'Lorem ipsum dolor sit amet, consectetur adipiscing elit, sed do eiusmod tempor incididunt ut labore et dolore magna aliqua. Ut enim ad minim veniam, quis nostrud',
      completionText: 'Congratulations! You have completed Piazza Guru achievement'
    }
  },
  {
    id: 10,
    title: "That's the Spirit",
    ability: AchievementAbility.EXPLORATION,
    isTask: true,
    position: 5,
    prerequisiteIds: [],
    goalIds: [9],
    cardTileUrl:
      'https://source-academy-assets.s3-ap-southeast-1.amazonaws.com/achievement/card-tile/thats-the-spirit-tile.png',
    view: {
      canvasUrl:
        'https://source-academy-assets.s3-ap-southeast-1.amazonaws.com/achievement/canvas/sample-canvas.png',
      description:
        'Lorem ipsum dolor sit amet, consectetur adipiscing elit, sed do eiusmod tempor incididunt ut labore et dolore magna aliqua. Ut enim ad minim veniam, quis nostrud',
      completionText: "Congratulations! You have completed That's the Spirit achievement"
    }
  },
  {
    id: 11,
    title: 'Kool Kidz',
    ability: AchievementAbility.FLEX,
    isTask: true,
    position: 6,
    prerequisiteIds: [],
    goalIds: [],
    cardTileUrl:
      'https://source-academy-assets.s3-ap-southeast-1.amazonaws.com/achievement/card-tile/annotated-tile.png',
    view: {
      canvasUrl:
        'https://source-academy-assets.s3-ap-southeast-1.amazonaws.com/achievement/canvas/sample-canvas.png',
      description: 'Diz for teh K00L K1dz',
      completionText: 'U SO G00D'
    }
  }
];

export const mockGoals: AchievementGoal[] = [
  {
    id: 0,
    text: 'Bonus for completing Colorful Carpet & Beyond the Second Dimension Achievements',
    maxExp: 100,
    type: GoalType.BINARY,
    meta: {
      condition: `AND(
        {
          event: 'achievement',
          restriction: 2,
        },
        {
          event: 'achievement',
          restriction: 1,
        },
      )`
    },
    exp: 0,
    completed: false
  },
  {
    id: 1,
    text: 'XP earned from Beyond the Second Dimension Mission',
    maxExp: 250,
    type: GoalType.ASSESSMENT,
    meta: { assessmentId: 5, requiredCompletionExp: 200 },
    exp: 213,
    completed: true
  },
  {
    id: 2,
    text: 'XP earned from Colorful Carpet Mission',
    maxExp: 250,
    type: GoalType.ASSESSMENT,
    meta: { assessmentId: 3, requiredCompletionExp: 200 },
    exp: 0,
    completed: false
  },
  {
    id: 3,
    text: 'Bonus for completing Curve Introduction & Curve Manipulation Achievements',
    maxExp: 100,
    type: GoalType.BINARY,
    meta: {
      condition: `AND(
      {
        event: 'achievement',
        restriction: 4,
      },
      {
        event: 'achievement',
        restriction: 5,
      },
    )`
    },
    exp: 0,
    completed: false
  },
  {
    id: 4,
    text: 'XP earned from Curve Introduction Mission',
    maxExp: 250,
    type: GoalType.ASSESSMENT,
    meta: { assessmentId: 7, requiredCompletionExp: 150 },
    exp: 178,
    completed: true
  },
  {
    id: 5,
    text: 'XP earned from Curve Manipulation Mission',
    maxExp: 250,
    type: GoalType.ASSESSMENT,
    meta: { assessmentId: 8, requiredCompletionExp: 200 },
    exp: 191,
    completed: false
  },
  {
    id: 6,
    text: 'Submit Source 3 Path',
    maxExp: 100,
    type: GoalType.BINARY,
    meta: {
      condition: `
      {
        event: 'assessment-submission',
        restriction: paths.SOURCE_THREE,
      },
    `
    },
    exp: 100,
    completed: true
  },
  {
    id: 7,
    text: 'XP earned from Source 3 Path',
    maxExp: 300,
    type: GoalType.ASSESSMENT,
    meta: {
      assessmentId: 12,
      requiredCompletionExp: 300
    },
    exp: 300,
    completed: true
  },
  {
    id: 8,
    text: 'Each Top Voted answer in Piazza gives 10 XP',
    maxExp: 100,
    type: GoalType.MANUAL,
    meta: {},
    exp: 40,
    completed: false
  },
  {
    id: 9,
    text: 'Submit 1 PR to Source Academy Github',
    maxExp: 100,
    type: GoalType.MANUAL,
    meta: {},
    exp: 100,
    completed: true
  },
  {
    id: 10,
    text: 'Be the Koolest Kidz in SOC by redeeming this 100 XP achievement yourself',
    maxExp: 100,
    type: GoalType.BINARY,
    meta: {
      condition: `
      {
        event: 'achievement',
        restriction: 'false',
      },
    `
    },
    exp: 0,
    completed: false
  }
];
