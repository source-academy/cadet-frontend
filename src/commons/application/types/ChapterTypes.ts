import { Variant } from 'js-slang/dist/types';

export type Chapter = {
    chapter: number;
    variant: Variant;
    displayName: string;
};