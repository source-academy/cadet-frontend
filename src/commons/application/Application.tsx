import { decompressFromEncodedURIComponent } from 'lz-string';
import * as qs from 'query-string';
import * as React from 'react';
import { Redirect, Route, RouteComponentProps, Switch } from 'react-router';

import { Variant } from 'js-slang/dist/types';

import Academy from '../../pages/academy/AcademyContainer';
import Contributors from '../../pages/contributors/Contributors';
import Login from '../../pages/login/LoginContainer';
import MissionControlContainer from '../../pages/missionControl/MissionControlContainer';
import NotFound from '../../pages/notFound/NotFound';
import Playground from '../../pages/playground/PlaygroundContainer';
import SourcecastContainer from '../../pages/sourcecast/SourcecastContainer';
import NavigationBar from '../navigationBar/NavigationBar';
import { stringParamToInt } from '../utils/ParamParseHelper';
import { Role, sourceLanguages } from './ApplicationTypes';
import { ExternalLibraryName, ExternalLibraryNames } from './types/ExternalTypes';

import Material from '../../containers/material/MaterialContainer'; // TODO: Remove

export type ApplicationProps = DispatchProps & StateProps & RouteComponentProps<{}>;

export type DispatchProps = {
  handleClearContext: (
    chapter: number,
    variant: Variant,
    externalLibraryName: ExternalLibraryName
  ) => void;
  handleEditorValueChange: (val: string) => void;
  handleEditorUpdateBreakpoints: (breakpoints: string[]) => void;
  handleEnsureLibrariesLoaded: () => void;
  handleLogOut: () => void;
  handleExternalLibrarySelect: (external: ExternalLibraryName) => void;
  handleSetExecTime: (execTime: string) => void;
};

export type StateProps = {
  accessToken?: string;
  currentPlaygroundChapter: number;
  currentPlaygroundVariant: Variant;
  role?: Role;
  title: string;
  name?: string;
  currentExternalLibrary: ExternalLibraryName;
};

const assessmentRegExp = ':assessmentId(-?\\d+)?/:questionId(\\d+)?';

class Application extends React.Component<ApplicationProps, {}> {
  public componentDidMount() {
    parsePlayground(this.props);
  }

  public render() {
    return (
      <div className="Application">
        <NavigationBar
          handleLogOut={this.props.handleLogOut}
          role={this.props.role}
          name={this.props.name}
          title={this.props.title}
        />
        <div className="Application__main">
          <Switch>
            <Route path="/academy" component={toAcademy(this.props)} />
            <Route path={`/mission-control/${assessmentRegExp}`} render={toIncubator} />
            <Route path="/playground" component={Playground} />
            <Route path="/login" render={toLogin(this.props)} />
            <Route path="/contributors" component={Contributors} />
            <Route path="/material" component={Material} />
            <Route path="/sourcecast" component={SourcecastContainer} />
            <Route exact={true} path="/" render={this.redirectToPlayground} />
            <Route component={NotFound} />
          </Switch>
        </div>
      </div>
    );
  }

  private redirectToPlayground = () => <Redirect to="/playground" />;
}

/**
 * A user routes to /academy,
 *  1. If the user is logged in, render the Academy component
 *  2. If the user is not logged in, redirect to /login
 */
const toAcademy = (props: ApplicationProps) =>
  props.accessToken === undefined || props.role === undefined
    ? () => <Redirect to="/login" />
    : () => <Academy accessToken={props.accessToken} role={props.role!} />;

const toLogin = (props: ApplicationProps) => () => (
  <Login luminusCode={qs.parse(props.location.search).code} />
);

const parsePlayground = (props: ApplicationProps) => {
  const prgrm = parsePrgrm(props);
  const chapter = parseChapter(props) || props.currentPlaygroundChapter;
  const variant = parseVariant(props, chapter) || props.currentPlaygroundVariant;
  const externalLibraryName = parseExternalLibrary(props) || props.currentExternalLibrary;
  const execTime = parseExecTime(props);
  if (prgrm) {
    props.handleEditorValueChange(prgrm);
    props.handleEnsureLibrariesLoaded();
    props.handleClearContext(chapter, variant, externalLibraryName);
    props.handleExternalLibrarySelect(externalLibraryName);
    props.handleSetExecTime(execTime);
  }
};

const toIncubator = (routerProps: RouteComponentProps<any>) => <MissionControlContainer />;

const parsePrgrm = (props: RouteComponentProps<{}>) => {
  const qsParsed = qs.parse(props.location.hash);
  // legacy support
  const program = qsParsed.lz !== undefined ? qsParsed.lz : qsParsed.prgrm;
  return program !== undefined ? decompressFromEncodedURIComponent(program) : undefined;
};

const parseChapter = (props: RouteComponentProps<{}>) => {
  const chapQuery = qs.parse(props.location.hash).chap;
  // find a language with this chapter (if any)
  const langWithMatchingChapter = sourceLanguages.find(language => language.chapter === chapQuery);

  const chap: number = langWithMatchingChapter
    ? langWithMatchingChapter.chapter
    : chapQuery === undefined
    ? NaN
    : parseInt(chapQuery, 10);

  return chap ? chap : undefined;
};

const parseVariant = (props: RouteComponentProps<{}>, chap: number) => {
  const variantQuery = qs.parse(props.location.hash).variant;
  // find a language with this variant and chapter (if any)
  const matchingLang = sourceLanguages.find(
    language => language.chapter === chap && language.variant === variantQuery
  );

  const variant: Variant = matchingLang ? matchingLang.variant : 'default';

  return variant;
};

const parseExternalLibrary = (props: RouteComponentProps<{}>) => {
  const ext = qs.parse(props.location.hash).ext || '';
  return Object.values(ExternalLibraryNames).includes(ext) ? ext : ExternalLibraryNames.NONE;
};

const parseExecTime = (props: RouteComponentProps<{}>) => {
  const time = qs.parse(props.location.hash).exec || '1000';
  // Parse the time string to a number, defaulting execTime to 1000
  const execTime = stringParamToInt(time) || 1000;
  return `${execTime < 1000 ? 1000 : execTime}`;
};

export default Application;
