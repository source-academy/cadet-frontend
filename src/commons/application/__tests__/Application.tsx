import { shallow } from 'enzyme';
import { Variant } from 'js-slang/dist/types';
import * as React from 'react';

import { mockRouterProps } from '../../mocks/ComponentMocks';
import Application, { ApplicationProps } from '../Application';
import { ExternalLibraryName } from '../types/ExternalTypes';

test('Application renders correctly', () => {
  const props: ApplicationProps = {
    ...mockRouterProps('/academy', {}),
    title: 'Cadet',
    currentPlaygroundChapter: 2,
    currentPlaygroundVariant: 'default',
    handleLogOut: () => {},
    currentExternalLibrary: ExternalLibraryName.NONE,
    handleClearContext: (
      chapter: number,
      variant: Variant,
      externalLibraryName: ExternalLibraryName
    ) => {},
    handleEditorValueChange: (val: string) => {},
    handleEditorUpdateBreakpoints: (breakpoints: string[]) => {},
    handleEnsureLibrariesLoaded: () => {},
    handleExternalLibrarySelect: (externalLibraryName: ExternalLibraryName) => {},
    handleSetExecTime: (execTime: string) => {}
  };
  const app = <Application {...props} />;
  const tree = shallow(app);
  expect(tree.debug()).toMatchSnapshot();
});
