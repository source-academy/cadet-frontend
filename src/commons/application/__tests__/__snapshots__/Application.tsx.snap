// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Application renders correctly 1`] = `
"<div className=\\"Application\\">
  <NavigationBar handleLogOut={[Function: handleLogOut]} role={[undefined]} name={[undefined]} title=\\"Cadet\\" />
  <div className=\\"Application__main\\">
    <Switch>
      <Route path=\\"/playground\\" component={[Function: C] { displayName: 'withRouter(Connect(Playground))', WrappedComponent: [Function: Playground], propTypes: { wrappedComponentRef: [Function: bound checkType] { isRequired: [Function: bound checkType] } } }} />
      <Route path=\\"/contributors\\" component={[function]} />
      <Route path=\\"/sourcecast/:sourcecastId?\\" component={[Function: C] { displayName: 'withRouter(Connect(Sourcecast))', WrappedComponent: [Function: Sourcecast], propTypes: { wrappedComponentRef: [Function: bound checkType] { isRequired: [Function: bound checkType] } } }} />
      <Route path=\\"/callback/github\\" component={[Function: GitHubCallback]} />
      <Route path=\\"/academy\\" render={[Function: redirectToLogin]} />
      <Route path=\\"/mission-control/:assessmentId(-?\\\\\\\\d+)?/:questionId(\\\\\\\\d+)?\\" render={[Function: toIncubator]} />
      <Route path=\\"/achievement\\" render={[Function: redirectToLogin]} />
      <Route path=\\"/login\\" render={[Function (anonymous)]} />
      <Route exact={true} path=\\"/\\" render={[Function: redirectToAcademy]} />
      <Route component={[Function: NotFound]} />
    </Switch>
  </div>
</div>"
`;

exports[`Application shows disabled when in disabled period 1`] = `
"<div className=\\"Application\\">
  <NavigationBar handleLogOut={[Function: handleLogOut]} role={[undefined]} name={[undefined]} title=\\"Cadet\\" />
  <div className=\\"Application__main\\">
    <Switch>
      <Route path=\\"/login\\" render={[Function (anonymous)]} />
      <Route path=\\"/academy\\" render={[Function: redirectToLogin]} />
      <Route exact={true} path=\\"/\\" render={[Function: redirectToLogin]} />
      <Route render={[Function: renderDisabled]} />
    </Switch>
  </div>
</div>"
`;
